opal/Makefile.am

```diff
@@ -53,8 +53,17 @@ lib@OPAL_LIB_PREFIX@open_pal_la_LIBADD = \
         util/libopalutil.la \
 	$(MCA_opal_FRAMEWORK_LIBS)
 lib@OPAL_LIB_PREFIX@open_pal_la_DEPENDENCIES = $(lib@OPAL_LIB_PREFIX@open_pal_la_LIBADD)
+
+if HAVE_ATTRIBUTE_DESTRUCTOR
+
 lib@OPAL_LIB_PREFIX@open_pal_la_LDFLAGS = -version-info $(libopen_pal_so_version)
 
+else
+
+lib@OPAL_LIB_PREFIX@open_pal_la_LDFLAGS = -version-info $(libopen_pal_so_version) -Wl,-fini,opal_cleanup
```

Comment:
I think you should treat this separately from whether the compiler supports attribute destructor.

I.e. you should be testing whether the linker supports the `-fini` option in configure.  And it should only be added to LDFLAGS here if the compiler does not support attribute destructor *and* the linker supports `-fini`.

If the compiler does not support attribute destructor and the linker does not support `-fini`, then we'll just have to leak memory / refuse to re-init / `opal_show_help("your system sucks; we can't re-init")` / whatever.